### DISCUSSION TITLE
Struggling with NextJS 13 and Vercel

### DISCUSSION THREAD

**Discussion by finom** [vercel/community#2979] [Bug Report]
I'm a big fan of NextJS, but some of the recent edge features have been challenging to work with.

- Middleware functionality is limited at the edge runtime, which means standard NodeJS functions, including essential modules like crypto, are not supported. This makes it difficult to check requests for authorization.
- ~Response streaming is not supported on Vercel, which is crucial for my OpenAI projects. As a result, I've had to move back to Heroku to ensure the necessary functionality.~
- The route.ts architecture requires multiple folders for a single route.ts file that only exports HTTP methods. To avoid this, I've had to combine NextJS with NestJS in a monorepo, which complicates the project. It also means needing developers with specific knowledge of NestJS concepts. It would be helpful to have a way to create fewer files and use nested API endpoints in a single something.ts file instead of route.ts. For instance, a small chat app with 4 API endpoints could be defined in one NestJS controller, but for larger apps, navigating and fixing bugs can become quite challenging.

<img width="177" alt="image" src="https://github.com/vercel/community/assets/1082083/c9e3fa70-0c80-4029-8a63-1de50453537e">

These issues have been a bit frustrating. While the app router was announced as a ready-to-go framework, it seems it was tested on a limited set of features. NextJS is excellent for large front-end applications with small REST APIs, but when additional functionality is needed, these restrictions can be quite challenging to work around. I hope future updates can address some of these concerns.

**Comment by leerob** [comment]
Hey @finom, I'm sorry you've had a bad experience here. A few notes:

1. Some crypto APIs are supported, see: https://nextjs.org/docs/app/api-reference/edge#crypto-apis
2. Response streaming should work, and is much easier now with our AI SDK, check this out: https://sdk.vercel.ai/docs/getting-started#create-an-api-route
3. I hear your feedback on the `route` convention. We typically are finding that routes will skip using a Route Handler and calling their `async` code directly from their Server Component (e.g. `page.tsx`), eliminating the need for that file. 

You are correct that right now, Next.js isn't designed to be for only REST APIs. While it does work, more powerful API frameworks like Nest might serve those needs better. Hope this helps 🙏

**Reply by finom** [reply]
@leerob thanks for the notes.

**Comment by armouti** [comment]
Agreed, it's growing into this behemoth of a mess. 

It was the perfect amount of abstraction on top of react when it was simple and focused. It should've stopped there, other stuff should've been separate libraries.

**Comment by finom** [comment]
Wow, I was extremely frustrated, but since then, a lot has happened. Just two weeks after my initial angry post, I started developing [my own solution for complex back-end apps](https://github.com/finom/vovk), which addressed the problem of having too many files.

Streaming on Vercel now works, but it requires a paid plan for LLMs with responses longer than 10 seconds. Although middlewares are used for only a few specific use cases, they are functional. The Edge Runtime restrictions have been mitigated by the community, which now creates platform-agnostic libraries.

Overall, I found solutions or workarounds for every issue I encountered. To gain more control over the back-end and avoid Vercel's restrictions, I am gradually transitioning to [Coolify](https://coolify.io/docs/videos), which is a great alternative to Vercel in some cases.

**Reply by leerob** [reply]
Hey @finom, glad to hear you've enjoyed your own customer backend solution.

By the way, the free plan on Vercel can stream responses for up to [60 seconds](https://vercel.com/changelog/vercel-functions-for-hobby-can-now-run-up-to-60-seconds) with the Node.js runtime.

**Reply by finom** [reply]
@leerob Thanks for the information. I apologize for my silly post; I was in a terrible state of mind due to the war. I should have phrased it differently.

**Reply by leerob** [reply]
It's okay, no hard feelings. I appreciate your feedback 😄

**Comment by mweel1** [comment]
Agreed.  They should stop pouring money into influencers and actually support their product.

**Reply by leerob** [reply]
Hey @mweel1, I'm sorry to hear you feel that way. For transparency, we have actually only ever sponsored one person, which is YouTube videos on Theo's channel. We care very deeply about supporting our product and are open to feedback on what we could do better.

[a comment by an automated account is not shown]
